### PR TITLE
added typing and export types in the package

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

Added types to pytest and add marker file for mypy

There a still a quite some type errors in here. 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] ~~Attached issue to pull request~~
- [x] Changelog entry
- [x] Code is clear and sufficiently documented
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] ~~End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~

### CODE REVIEW

**Review by bartv — APPROVED** [review]
Very nice!

**Review by edvgui — APPROVED** [review]
I like that a lot

File: pytest_inmanta/plugin.py [hunk 1]
```diff
@@ -185,7 +197,9 @@ def project_shared(project_factory):
 
 # Temporary workaround for plugins loading multiple times (inmanta/pytest-inmanta#49)
 @pytest.fixture(scope="session")
-def project_shared_no_plugins(project_factory):
+def project_shared_no_plugins(
+    project_factory: typing.Iterator[typing.Callable[[], "Project"]]
```
Comment (arnaudsjs) [inline pytest_inmanta/plugin.py:201]:
```suggestion
    project_factory: typing.Callable[[Dict[str, Any]], "Project"]
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: pytest_inmanta/plugin.py [hunk 2]
```diff
@@ -384,59 +400,64 @@ class Project:
...
-    def init(self, capsys):
+    def init(self, capsys: CaptureFixture) -> None:
         self._stdout = None
         self._stderr = None
         self._capsys = capsys
         self.types = None
         self.version = None
         self.resources = {}
-        self._root_scope = {}
+        self._root_scope = None
```
Comment (arnaudsjs) [inline pytest_inmanta/plugin.py:432]: This is a breaking change in the senses that `get_root_scope()` can return None now. I am not sure whether that's an issue.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: pytest_inmanta/plugin.py (hunk 2 above)
Comment (wouterdb) [reply pytest_inmanta/plugin.py:432]: The rootscope is a NameSpace object, so it doesn't really matter I think

File: pytest_inmanta/plugin.py [hunk 3]
```diff
@@ -773,14 +802,14 @@ def unittest_resource_exists(self, name: str) -> bool:
...
-    ) -> Dict[str, Union[str, bool, float, int]]:
+    ) -> typing.Dict[str, typing.Union[str, bool, float, int]]:
         """
         Get the state of the unittest resource
         """
         return DATA[name]
 
     def unittest_resource_set(
-        self, name: str, **kwargs: Union[str, bool, float, int]
+        self, name: str, **kwargs: typing.Union[str, bool, float, int]
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply pytest_inmanta/plugin.py:812]: for kwargs you have to give the type of the values, I think https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html

**Review by arnaudsjs — APPROVED** [review]

**Pull request merged**